pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -490,6 +526,29 @@ void ConsumerImpl::messageProcessed(Message& msg) {
     increaseAvailablePermits(currentCnx);
 }
 
+/**
+ * Clear the internal receiver queue and returns the message id of what was the 1st message in the queue that was
+ * not seen by the application
+ */
+Optional<BatchMessageId> ConsumerImpl::clearReceiveQueue() {
+    Message nextMessageInQueue;
+    if (incomingMessages_.peekAndClear(nextMessageInQueue)) {
+        // There was at least one message pending in the queue
+        // We can safely cast to 'BatchMessageId' since all the messages queued will have that type of message id,
+        // irrespective of whether they were part of a batch or not.
+        const MessageId& nextMessageId = static_cast<const BatchMessageId&>(nextMessageInQueue.getMessageId());
+        BatchMessageId previousMessageId(nextMessageId.ledgerId_, nextMessageId.entryId_ - 1);
```

Comment:
So here you cast the BatchMessageId to MessageId again (i.e ignore the batchIndex) hence stand a risk of redelivering the message once the batch is redelivered

On the other hand if you maintain the batchIndex then you can't clear the unAckedMessageTracter and bathIndexTracker on reconnection.

Reply:
That's actually a bug. Need to fix this